Add user management ability

cjhubert opened the pull request:
This PR adds the following:
- user_management recipe, which will require auth on mongo config
- user resource and provider
- default users attributes
- user_management integration tests

Should take care of #225

cjhubert committed:
adding the ability to add admin and other users to mongodb through recipe, readme updates, kitchen integration tests

cjhubert committed:
updating docs and test to work cross-distros

cjhubert committed:
removing superfluous auth config from default recipe

cjhubert committed:
updating the tests to be less crazy

cjhubert committed:
updating recipe / attribute so admin user data only needs to be overridden, and is concat'd onto users array if auth is true
updating read me and kitchen.yml for more instructions

cjhubert committed:
removing redundant code on add_user

cjhubert committed:
Merge pull request #1 from ceejh/add-user-management

adding the ability to add admin and other users to mongodb through recip...

josephholsten commented on providers/user.rb:
```diff
@@ -0,0 +1,70 @@
+  # this will fail on the first attempt, but user will still be created
+  # because of the localhost exception
+  begin
+    admin.authenticate(node[:mongodb][:admin][:username], node[:mongodb][:admin][:password])
```
this won't work with multiple mongos on the same box (I know nothing does, but I'd like to avoid having to change this later). Same goes for the `retrieve_db`'s access of `node[:mongodb][:host]`.

Perhaps a connection object like in [opscode's `database` cookbook](https://github.com/opscode-cookbooks/database)?

josephholsten commented:
looks awesome! If it's easy to extract out a connection object (preferably with the same defaults you provide), I'd love it. but it lgtm regardless.

cjhubert committed:
adding connection hash to resource, using it in provider and changing default to pass config

cjhubert committed:
Merge pull request #2 from ceejh/add-connection-hash

adding connection hash to resource, using it in provider and changing de...

cjhubert commented:
I've updated the code to factor out the connection object. Let me know if there are any problems with it.
Edit: Actually, looking again, I probably forgot to change all the `admin.authenticate`s and just changed `retrieve_db`. Let me know if I was on the right track and I'll update `authenticate` as well.

josephholsten commented on recipes/user_management.rb:
```diff
@@ -0,0 +1,20 @@
+
+# If authentication is required,
+# add the admin to the users array for adding/updating
+users = [admin] if node['mongodb']['config']['auth'] == true
```
`users` is undefined if `node['mongodb']['config']['auth'] != true`

this'll make the `concat` have a bad time.

cjhubert replied:
Nice catch, should now be fixed.

josephholsten commented on providers/user.rb:
```diff
@@ -0,0 +1,70 @@
+
+  Mongo::MongoClient.new(
+    @new_resource.connection[:host],
+    @new_resource.connection[:port],
```
I'm not sure these are the right attribute keys. Does this pass your kitchen test?

cjhubert replied:
It does pass all the kitchen tests. I logged onto the converged machines to double-verify that the state was as expected.

josephholsten commented:
@ceejh definitely the right track! Let me know when you're comfortable with it and I'll run a full kitchen pass against it.

Docs, tests, resources, works with multitenancy, this is quite the patch!

cjhubert committed:
fixing undefined users array if auth is false and admin.authenticate calls to use connection

cjhubert committed:
forgot to run rubocop, added an offense, fixed using array literal instead

cjhubert commented:
Should be good to go unless you see any more problems.

jamesonjlee commented on attributes/users.rb:
```diff
@@ -0,0 +1,10 @@
+default['mongodb']['config']['auth'] = true
+
+default[:mongodb][:admin] = {
```
`'` instead of `:` also goes for the arrays.

jamesonjlee commented:
other than the style nitpick LGTM

cjhubert committed:
using single-quote marks instead of symbols for attributes

cjhubert commented:
Updated the keys to use `'`

cjhubert commented:
Let me know if there's anything else that needs to be done to get this merged.

josephholsten commented on .kitchen.yml:
```diff
@@ -120,3 +120,18 @@ suites:
+      install_method: mongodb-org
+      # Needed to read the correct config file
+      # since mongo 2.6
+      default_init_name: mongod
```
:cry: We'll fix #262 soon.
